docs/ide/api-designs/Workspace and Source Generated Documents.md

```diff
@@ -0,0 +1,73 @@
+
+## APIs on `Project`
+
+### Existing API: `Documents`
+
+Note the Documents collection will _not_ change, and will only include regular documents. It cannot include source generated documents because since it is a property access, there is no opportunity to make this async, nor have any way to express cancellation. Since a generator could produce any number of documents, there's simply no way to make this answer quickly without fundamentally taking a different approach to the API.
+
+### New API: `ImmutableArray<SourceGeneratedDocument> GetSourceGeneratedDocumentsAsync(CancelationToken)`
+
+This will run generators if they have not already ran, and then return a `SourceGeneratedDocument` for each generated document.
+
+The implementation of this will run `GetCompilationAsync` if the documents have not been generated; if the Compilation had already been generated, we would have cached the generated document information and so this would be cheap. We will hold onto the list of documents strongly (the tree will be a recoverable tree), so even if the `Compilation` is GC'ed we won't have to recompute this part a second time.
+
+### New API: `SourceGeneratedDocument? GetSourceGeneratedDocumentAsync(DocumentId, CancellationToken)`
+
+Fetches a single document by ID; equivalent to calling the API and filtering down to the right document.
+
+### Existing API: `GetDocument(SyntaxTree)`
+
+No changes, it will potentially return a `SourceGeneratedDocument` now. Callers _may_ have to check whether it's a generated document before they try to offer a fix or refactoring on the document.
+
+This API is the hardest API to figure out what to do in this entire spec. If it doesn't return generated documents, many features would break, since it's very common to assume that if a syntax tree exists in a `Compilation`, that there must be a `Document` that matches it. However, some features might _not_ want to see a generated document returned here because then they're going to try to modify that, and that won't make sense either. An audit of the code fixes in Roslyn discovered that for many of the places that would need an update to _ignore_ a generated document would also need an update to deal with the null reference that would come back from `GetDocument`.
+
+The other odd thing here is whether this function needs to be asynchronous or not. The current belief is no, because the only way you can have gotten a SyntaxTree to pass to it from the same `Solution` is either from:
+
+1. The `Compilation` or a symbol that came from it, which means generators have ran.
+2. You inspected some `SourceGeneratedDocument` and got it's tree, which means that generator has already ran.
```

Comment:
i'm a bit confused by this though and some previous stuff you wrote.  Say the SourceGeneratedDoc is there and it's tree is avialable (either because it's in memory, or we rehydrated the recoerable tree for it).  That won't keep the compiltion around.  So if we recreate the compilat (and thus rerun source generators), what does hte hookup again?

Or is it the case that we will recreate the compilation and *not* run source generators and will *instead* just manually re-add the generated trees to it?

Reply:
> Or is it the case that we will recreate the compilation and not run source generators and will instead just manually re-add the generated trees to it?

This.